spec-0007/index.md

```diff
@@ -0,0 +1,149 @@
+- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.
+
+Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.
+
+### Scope
+
+This is intended as a recommendation to all libraries that allow users to control the
+state of a NumPy random number generator. It is specifically targeted toward functions
+that allow use of `numpy.random.seed` to control the random state and accept
+`RandomState` instances via an argument with a name other than `rng`, but the ideas
+are more broadly applicable. Use of random number generators other than those provided
+by NumPy are beyond the scope of this SPEC.
+
+### Concepts
+
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
+- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
+- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.
+
+## Implementation
+
+Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
+
+Two strong motivations for moving over to `Generator`s are:
+
+1. they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
+2. they avoid using global state (from `numpy.random.mtrand._rand`).
+
+Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen principle[^hinsen],
+although it could very nearly do so by enforcing the use of `rng` as a keyword argument.
+
+[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.
+
+The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:
+
+1. Accept both `rng` and `random_state`/`seed` keyword arguments.
+2. If both are specified, raise an error.
+3. If neither is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
+4. If `random_state`/`seed` is passed by keyword or by position, treat it as before, but:
+
+   - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.
```

Comment:
I suggest adding a recommendation here to wait N releases (with N determined from SPEC 0) before starting to emit a deprecation warning after introducing `rng=`, because downstream libraries cannot switch to the new keyword when they're still supporting library versions that don't have it.

Reply:
Would we specify a number of releases or number of years?

Reply:
SPEC 0 is in years, but projects work in releases. So the text could be amended with either option. Here is a concrete suggestion: _"After `rng=` is introduced and is available in all releases within the support window suggested by SPEC 0 for the package, emit a `DeprecationWarning` ...."_

Reply:
That I can do. I just couldn't specify `N`. Will do today.

Reply:
Addressed by https://github.com/stefanv/specs/pull/3/commits/6ec3b6ae1dd5ccc012eb16b5d2c218194c9120fa

Reply:
Thanks. The change LGTM.